src/Sylius/Component/Core/Promotion/Applicator/UnitsPromotionAdjustmentsApplicator.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Promotion\Applicator;
+
+use Sylius\Component\Core\Distributor\IntegerDistributorInterface;
+use Sylius\Component\Core\Model\AdjustmentInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemInterface;
+use Sylius\Component\Order\Factory\AdjustmentFactoryInterface;
+use Sylius\Component\Order\Model\OrderItemUnitInterface;
+use Sylius\Component\Originator\Originator\OriginatorInterface;
+use Sylius\Component\Promotion\Model\PromotionInterface;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class UnitsPromotionAdjustmentsApplicator implements UnitsPromotionAdjustmentsApplicatorInterface
+{
+    /**
+     * @var AdjustmentFactoryInterface
+     */
+    private $adjustmentFactory;
+
+    /**
+     * @var IntegerDistributorInterface
+     */
+    private $distributor;
+
+    /**
+     * @var OriginatorInterface
+     */
+    private $originator;
+
+    /**
+     * @param AdjustmentFactoryInterface $adjustmentFactory
+     * @param IntegerDistributorInterface $distributor
+     * @param OriginatorInterface $originator
+     */
+    public function __construct(
+        AdjustmentFactoryInterface $adjustmentFactory,
+        IntegerDistributorInterface $distributor,
+        OriginatorInterface $originator
+    ) {
+        $this->adjustmentFactory = $adjustmentFactory;
+        $this->distributor = $distributor;
+        $this->originator = $originator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(OrderInterface $order, PromotionInterface $promotion, array $adjustmentsAmounts)
+    {
+        if ($order->countItems() !== count($adjustmentsAmounts)) {
+            throw new \InvalidArgumentException(
+                'Number of adjustments amount to distribute must be equal with number of order items.'
+            );
+        }
+
+        $i = 0;
+        foreach ($order->getItems() as $item) {
+            if (0 === $adjustmentsAmounts[$i]) {
```

Comment:
Can we have `0` value in a middle of `adjustmentsAmounts` array? 
If yes, this code has a bug. For example for amounts: 1, 0, 1 and 3 items, this loop till apply adjustment to first item, but  when it will check the second element of array it will stop in this loop, because we don't increment index.

If no, break will be much more appropriate here.